### PR TITLE
Checkout

### PR DESCRIPTION
**Pull request by Ylambers**

### CODE REVIEW

File: src/Controller/CartController.php [hunk 1]
```diff
@@ -49,14 +59,24 @@ public function addLineItemAction(Request $request)
...
+            $cartId = $session->get(CartRepository::CART_ID);
             $country = \Locale::getRegion($locale);
             $currency = $this->getParameter('commercetools.currency.'. $country);
+
             /**
              * @var CartRepository $repository
              */
             $repository = $this->get('commercetools.repository.cart');
-            $repository->addLineItem($request->getLocale(), $cartId, $productId, $variantId, $quantity, $currency, $country);
+            $repository->addLineItem(
```
Comment (Haehnchen) [inline src/Controller/CartController.php]: pls refactor to use a class as argument ;)

File: src/Controller/CheckoutController.php [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+use Symfony\Component\HttpFoundation\Session\Session;
+use Symfony\Component\HttpKernel\EventListener\SessionListener;
+use Symfony\Component\Validator\Tests\Fixtures\Entity;
+
+class CheckoutController extends Controller
+{
+    public function signinAction(Request $request)
+    {
+        if ($this->get('security.authorization_checker')->isGranted('IS_AUTHENTICATED_FULLY')) {
+            return new RedirectResponse($this->generateUrl('_ctp_example_checkout_address'));
```
Comment (Haehnchen) [inline src/Controller/CheckoutController.php]: controller shortcut: `$this->redirect...`

File: src/Controller/CheckoutController.php [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+        $cartId = $session->get(CartRepository::CART_ID);
+        $cart = $this->get('commercetools.repository.cart')->getCart($request->getLocale(), $cartId, $customerId);
+
+
+        if (is_null($cart->getId())) {
+            return new RedirectResponse($this->generateUrl('_ctp_example_cart'));
+        }
+
+        $entity = CartEntity::ofCart($cart);
+        $form = $this->createFormBuilder($entity)
```
Comment (Haehnchen) [inline src/Controller/CheckoutController.php]: extract to FormType class; reusability

File: src/Controller/CheckoutController.php [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+
+            if (!is_null($cart)) {
+                return new RedirectResponse($this->generateUrl('_ctp_example_checkout_shipping'));
+            }
+        }
+
+        if (!$form->isSubmitted() && count(array_diff_key($cart->getShippingAddress()->toArray(), ['country' => true])) == 0 ) {
+            if (!is_null($customer)) {
+                $address = $customer->getDefaultShippingAddress();
+                $form->get('shippingAddress')->get('salutation')->setData($address->getSalutation());
```
Comment (Haehnchen) [inline src/Controller/CheckoutController.php]: `data_class` usage

File: src/Controller/UserController.php [hunk 5]
```diff
@@ -50,47 +63,26 @@ public function loginAction(Request $request)
...
 
-        $userDetails = $form->getData();
         if ($form->isValid() && $form->isSubmitted()){
 
-            $firstName = $form->getData()['firstName'];
-            $lastName = $form->getData()['lastName'];
-            $email = $form->getData()['email'];
-            $currentPassword = $form->getData()['currentPassword'];
-            $newPassword = $form->getData()['newPassword'];
+            $firstName = $form->get('firstName')->getData();
```
Comment (Haehnchen) [inline src/Controller/UserController.php]: use `entity`?, also stores data

**Comment by jayS-de** [comment]
Opened new PR to develop branch

**Pull request closed without merging**